projects/28_gltf/main.cpp

```diff
@@ -0,0 +1,837 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <functional>
+#include <utility>
+#include <queue>
+#include <unordered_set>
+
+#include "ppx/ppx.h"
+#include "ppx/timer.h"
+#include "ppx/camera.h"
+#include "ppx/graphics_util.h"
+#include "ppx/grfx/grfx_scope.h"
+#define CGLTF_IMPLEMENTATION
+#include "cgltf.h"
+#include "glm/gtc/type_ptr.hpp"
+
+using namespace ppx;
+
+#if defined(USE_DX11)
+const grfx::Api kApi = grfx::API_DX_11_1;
+#elif defined(USE_DX12)
+const grfx::Api kApi = grfx::API_DX_12_0;
+#elif defined(USE_VK)
+const grfx::Api kApi = grfx::API_VK_1_1;
+#endif
+
+class ProjApp
+    : public ppx::Application
+{
+public:
+    virtual void Config(ppx::ApplicationSettings& settings) override;
+    virtual void Setup() override;
+    virtual void Render() override;
+
+private:
+    struct PerFrame
+    {
+        grfx::CommandBufferPtr cmd;
+        grfx::SemaphorePtr     imageAcquiredSemaphore;
+        grfx::FencePtr         imageAcquiredFence;
+        grfx::SemaphorePtr     renderCompleteSemaphore;
+        grfx::FencePtr         renderCompleteFence;
+    };
+
+    struct Texture
+    {
+        grfx::ImagePtr            pImage;
+        grfx::SampledImageViewPtr pTexture;
+        grfx::SamplerPtr          pSampler;
+    };
+
+    struct Material
+    {
+        grfx::PipelineInterfacePtr pInterface;
+        grfx::GraphicsPipelinePtr  pPipeline;
+        grfx::DescriptorSetPtr     pDescriptorSet;
+        std::vector<Texture>       textures;
+    };
+
+    struct Primitive
+    {
+        grfx::Mesh* mesh;
+    };
+
+    struct Renderable
+    {
+        Material*            pMaterial;
+        Primitive*           pPrimitive;
+        grfx::DescriptorSet* pDescriptorSet;
+
+        Renderable(Material* m, Primitive* p, grfx::DescriptorSet* set)
+            : pMaterial(m), pPrimitive(p), pDescriptorSet(set) {}
+    };
+
+    struct Object
+    {
+        float4x4                model;
+        grfx::BufferPtr         pUniformBuffer;
+        std::vector<Renderable> renderables;
+    };
+
+    using RenderList   = std::unordered_map<Material*, std::vector<Object*>>;
+    using TextureCache = std::unordered_map<std::string, grfx::ImagePtr>;
+
+    std::vector<PerFrame>        mPerFrame;
+    grfx::DescriptorPoolPtr      mDescriptorPool;
+    grfx::DescriptorSetLayoutPtr mSetLayout;
+    grfx::ShaderModulePtr        mVertexShader;
+    grfx::ShaderModulePtr        mPixelShader;
+    PerspCamera                  mCamera;
+    float3                       mLightPosition = float3(10, 100, 10);
+
+    RenderList             mRenderList;
```

Comment:
This is never used. I suppose you wanted to use this to sort by material and reduce the state switching?

Reply:
Oh yes, thanks, yes iterated a bit around this, forgot to clean this up.